integration/fuse/src/main/java/alluxio/fuse/file/FuseFileOutStream.java

```diff
@@ -0,0 +1,175 @@
+   *
+   * @param fileSystem the Alluxio file system
+   * @param authPolicy the Authentication policy
+   * @param uri the alluxio uri
+   * @param flags the fuse create/open flags
+   * @param mode the filesystem mode, -1 if not set
+   * @param status the uri status
+   * @return a {@link FuseFileInOrOutStream}
+   */
+  public static FuseFileOutStream create(FileSystem fileSystem, AuthPolicy authPolicy,
+      AlluxioURI uri, int flags, long mode, Optional<URIStatus> status) {
+    Preconditions.checkNotNull(fileSystem);
+    Preconditions.checkNotNull(authPolicy);
+    Preconditions.checkNotNull(uri);
+    Preconditions.checkNotNull(status);
+    if (mode == AlluxioFuseUtils.MODE_NOT_SET_VALUE && status.isPresent()) {
+      mode = status.get().getMode();
+    }
+    long fileLen = status.map(URIStatus::getLength).orElse(0L);
+    if (status.isPresent()) {
+      if (AlluxioFuseOpenUtils.containsTruncate(flags)) {
+        AlluxioFuseUtils.deleteFile(fileSystem, uri);
+        fileLen = 0;
+        LOG.debug(String.format("Open path %s with flag 0x%x for overwriting. "
+            + "Alluxio deleted the old file and created a new file for writing", uri, flags));
+      } else {
+        // Support open(O_WRONLY flag) - truncate(0) - write() workflow, otherwise error out
+        return new FuseFileOutStream(fileSystem, authPolicy, Optional.empty(), fileLen, uri, mode);
+      }
+    }
+    return new FuseFileOutStream(fileSystem, authPolicy,
+        Optional.of(AlluxioFuseUtils.createFile(fileSystem, authPolicy, uri, mode)),
+        fileLen, uri, mode);
+  }
+
+  private FuseFileOutStream(FileSystem fileSystem, AuthPolicy authPolicy,
+      Optional<FileOutStream> outStream, long fileLen, AlluxioURI uri, long mode) {
+    Preconditions.checkNotNull(fileSystem);
+    Preconditions.checkNotNull(authPolicy);
+    Preconditions.checkNotNull(outStream);
+    Preconditions.checkNotNull(uri);
+    mFileSystem = fileSystem;
+    mAuthPolicy = authPolicy;
+    mOutStream = outStream;
+    mOriginalFileLen = fileLen;
+    mURI = uri;
+    mMode = mode;
+  }
+
+  @Override
+  public int read(ByteBuffer buf, long size, long offset) {
+    throw new UnsupportedOperationException("Cannot read from write only stream");
+  }
+
+  @Override
+  public synchronized void write(ByteBuffer buf, long size, long offset) {
+    if (size > Integer.MAX_VALUE) {
+      throw new UnsupportedOperationException(
+          String.format("Cannot write more than %s", Integer.MAX_VALUE));
+    }
+    if (!mOutStream.isPresent()) {
+      throw new UnsupportedOperationException(
+          "Cannot overwrite/extending existing file without O_TRUNC flag or truncate(0) operation");
+    }
+    int sz = (int) size;
+    long bytesWritten = mOutStream.get().getBytesWritten();
+    if (offset != bytesWritten && offset + sz > bytesWritten) {
+      throw new UnsupportedOperationException(String.format("Only sequential write is supported. "
```

Comment:
This is where we should handle the out of order write in a certain range, eg: 1M, But it is quite strange we didn't hit this issue in our tests.

Reply:
Yeah we didn't hit it in dd test, but if FIO can hit it, we can easily reproduce it with FIO?